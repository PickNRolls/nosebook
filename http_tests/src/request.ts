import supertest, { Test } from 'supertest';
import TestAgent from 'supertest/lib/agent';

const HOST = 'http://backend:8080';
const SESSION_HEADER_KEY = 'X-Auth-Session-Id';

class TestAgentWrapper {
  private host: string;
  private agent: TestAgent;

  public constructor(host: string) {
    this.host = host;
    this.agent = supertest(host);
  }

  public unwrap(): TestAgent {
    return this.agent;
  }

  public extend(prefixUrl: string): TestAgentWrapper {
    return new TestAgentWrapper(this.host + prefixUrl);
  }

  private hook<M extends 'get' | 'post'>(method: M, url: string): Test {
    const test = this.agent[method](url);

    test.set(SESSION_HEADER_KEY, 'bb23af03-be50-4bce-b729-b259b2e02e54');

    return test;
  }

  public get(url: string): Test {
    return this.hook('get', url);
  }

  public post(url: string): Test {
    return this.hook('post', url);
  }
};

export default new TestAgentWrapper(HOST);


