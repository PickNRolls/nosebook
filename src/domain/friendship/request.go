package domainfriendship

import (
	"nosebook/src/errors"
	"time"

	"github.com/google/uuid"
)

type FriendRequest struct {
	RequesterId uuid.UUID `json:"requesterId" db:"requester_id"`
	ResponderId uuid.UUID `json:"responderId" db:"responder_id"`
	Message     string    `json:"message" db:"message"`
	Accepted    bool      `json:"accepted" db:"accepted"`
	Viewed      bool      `json:"viewed" db:"viewed"`
	CreatedAt   time.Time `json:"createdAt" db:"created_at"`

	events []Event
}

func New(
	requesterId uuid.UUID,
	responderId uuid.UUID,
	message string,
	accepted bool,
	viewed bool,
	createdAt time.Time,
	raiseCreatedEvent bool,
) *FriendRequest {
	output := &FriendRequest{
		RequesterId: requesterId,
		ResponderId: responderId,
		Message:     message,
		Accepted:    accepted,
		Viewed:      viewed,
		CreatedAt:   createdAt,
		events:      make([]Event, 0),
	}

	if raiseCreatedEvent {
		output.raiseEvent(NewCreatedEvent())
	}

	return output
}

func (this *FriendRequest) raiseEvent(event Event) {
	this.events = append(this.events, event)
}

func (this *FriendRequest) Events() []Event {
	return this.events
}

func (this *FriendRequest) ViewBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отметить заявку просмотренной")
	}

	this.Viewed = true
	this.raiseEvent(NewViewedEvent())

	return nil
}

func (this *FriendRequest) AcceptBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может принять заявку")
	}

	this.Accepted = true
	this.Viewed = true
	this.raiseEvent(NewAcceptedEvent())

	return nil
}

func (this *FriendRequest) DenyBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отклонить заявку")
	}

	this.Accepted = false
	this.Viewed = true
	this.raiseEvent(NewDeniedEvent())

	return nil
}

func (this *FriendRequest) RemoveBy(userId uuid.UUID) *errors.Error {
	if !this.Accepted {
		return errors.New("Friendship Error", "Нельзя удалить непринятую заявку")
	}

	this.Accepted = false
	previousRequesterId := this.RequesterId
	previousResponderId := this.ResponderId

	if this.RequesterId == userId {
		this.RequesterId, this.ResponderId = this.ResponderId, this.RequesterId
	}

	this.raiseEvent(NewRemovedEvent(previousRequesterId, previousResponderId))

	return nil
}
