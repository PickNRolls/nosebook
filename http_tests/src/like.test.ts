import request from './request';

describe('like', () => {
  let like = request.extend({
    prefixUrl: '/like'
  });

  test('POST /post', async () => {
    async function doLike() {
      return await like
        .post('/post')
        .send({
          id: 'c7b7bf17-38f9-4ed5-b0a8-501a90f7c829',
        })
        .expect(200);
    }

    expect((await doLike()).body).toMatchSnapshot();
    expect((await doLike()).body).toMatchSnapshot();
  });

  test('POST /comment', async () => {
    async function doLike() {
      return await like
        .post('/comment')
        .send({
          id: '620c79b7-3927-48b7-a308-1ffd3db6036f',
        })
        .expect(200);
    }

    expect((await doLike()).body).toMatchSnapshot();
    expect((await doLike()).body).toMatchSnapshot();
  });
});

