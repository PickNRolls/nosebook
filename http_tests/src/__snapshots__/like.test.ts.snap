// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`like POST /comment 1`] = `
{
  "data": {
    "commentId": "620c79b7-3927-48b7-a308-1ffd3db6036f",
    "liked": true,
  },
  "ok": true,
}
`;

exports[`like POST /comment 2`] = `
{
  "data": {
    "commentId": "620c79b7-3927-48b7-a308-1ffd3db6036f",
    "liked": false,
  },
  "ok": true,
}
`;

exports[`like POST /post 1`] = `
{
  "data": {
    "liked": true,
    "postId": "c7b7bf17-38f9-4ed5-b0a8-501a90f7c829",
  },
  "ok": true,
}
`;

exports[`like POST /post 2`] = `
{
  "data": {
    "liked": false,
    "postId": "c7b7bf17-38f9-4ed5-b0a8-501a90f7c829",
  },
  "ok": true,
}
`;
